extension/awsmiddleware/middleware.go

```diff
@@ -0,0 +1,182 @@
+	// ID must be unique. It cannot clash with existing middleware.
+	ID() string
+	// Position to insert the handler.
+	Position() HandlerPosition
+}
+
+// RequestHandler allows for custom processing of requests.
+type RequestHandler interface {
+	metadata
+	HandleRequest(r *http.Request)
+}
+
+// ResponseHandler allows for custom processing of responses.
+type ResponseHandler interface {
+	metadata
+	HandleResponse(r *http.Response)
+}
+
+// Middleware defines the request and response handlers to be configured
+// on AWS Clients.
+type Middleware interface {
+	RequestHandlers() []RequestHandler
+	ResponseHandlers() []ResponseHandler
+}
+
+// Extension is an extension that implements Middleware.
+type Extension interface {
+	extension.Extension
+	Middleware
+}
+
+// ConfigureSDKv1 adds middleware to the AWS SDK v1. Request handlers are added to the
+// Build handler list and response handlers are added to the Unmarshal handler list.
+func ConfigureSDKv1(mw Middleware, handlers *request.Handlers) error {
+	var errs error
+	for _, handler := range mw.RequestHandlers() {
+		if err := addHandlerToList(&handlers.Build, namedRequestHandler(handler), handler.Position()); err != nil {
+			errs = errors.Join(errs, fmt.Errorf("%w (%q): %w", errInvalidHandler, handler.ID(), err))
+		}
+	}
+	for _, handler := range mw.ResponseHandlers() {
+		if err := addHandlerToList(&handlers.Unmarshal, namedResponseHandler(handler), handler.Position()); err != nil {
+			errs = errors.Join(errs, fmt.Errorf("%w (%q): %w", errInvalidHandler, handler.ID(), err))
+		}
+	}
+	return errs
+}
+
+// ConfigureSDKv2 adds middleware to the AWS SDK v2. Request handlers are added to the
+// Build step and response handlers are added to the Deserialize step.
+func ConfigureSDKv2(mw Middleware, config *aws.Config) error {
+	var errs error
+	for _, handler := range mw.RequestHandlers() {
+		relativePosition, err := toRelativePosition(handler.Position())
+		if err != nil {
+			errs = errors.Join(errs, fmt.Errorf("%w (%q): %w", errInvalidHandler, handler.ID(), err))
+			continue
+		}
+		rmw := &requestMiddleware{RequestHandler: handler, position: relativePosition}
+		config.APIOptions = append(config.APIOptions, func(stack *middleware.Stack) error {
+			return stack.Build.Add(rmw, rmw.position)
+		})
+	}
+	for _, handler := range mw.ResponseHandlers() {
+		relativePosition, err := toRelativePosition(handler.Position())
+		if err != nil {
+			errs = errors.Join(errs, fmt.Errorf("%w (%q): %w", errInvalidHandler, handler.ID(), err))
+			continue
+		}
+		rmw := &responseMiddleware{ResponseHandler: handler, position: relativePosition}
+		config.APIOptions = append(config.APIOptions, func(stack *middleware.Stack) error {
+			return stack.Deserialize.Add(rmw, rmw.position)
+		})
+	}
+	return errs
+}
+
+// addHandlerToList adds the handler to the list based on the position.
+func addHandlerToList(handlerList *request.HandlerList, handler request.NamedHandler, position HandlerPosition) error {
```

Comment:
super duper mild nit - I personally dislike the [verb] to [thing] naming convention, but that is purely personal preference. It would be nice if we could just use `append` but that's already reserved for slices :/ oh well